More work towards Custom Indexes

Maxxen opened the pull request:
This PR does the following:

- Adds a new IndexType catalog entry, with a name and function pointer to instantiate a new index instance
- Adds a new "UnknownIndex" index subclass, working as a placeholder so that the system is aware of index-types pushed by extensions that are not or are yet to be loaded (e.g. during WAL playback). They also hold on to the related index creation data from the WAL, which we can pass on to the extension index constructor to lazily initialize it.
- Moves the decision of whether or not an index scan can be performed into the index class itself. Right now we go with the first one that succeeds but in the future we might want to do some cardinality/cost estimation or go with the most specific match.
- Also does some more decoupling between the system and the ART so that custom indexes can be used instead.

There's still a lot more to do:
- Handle undo buffer/transactions for custom indexes
- Handle constraints for custom indexes.
- Handle logical/physical plan creation for custom indexes
- Handle creation of custom indexes through SQL if the index type entry is available.
- Handle registration of custom indexes in the ExtensionUtil class.

I also want to rework how index scans are handled so that they can be implemented by index types more like table functions. (Maybe even make it possible to support "covering" indexes too?). Right now it seems like we won't ever index scan more than SVS?

Nevertheless I figured it would be good to "checkpoint" this work in this PR to make it more manageable to review and as it already allows for some (very) limited custom index functionality in extensions by using optimizer rules and plan replacements.

Maxxen committed:
resolve merge

Maxxen committed:
rebase

Maxxen committed:
add index-type catalog entry, art as default index

Maxxen committed:
default to ART

Maxxen committed:
make indextype creation in schema optional so external catalogs dont have to worry about it

Maxxen committed:
lazily load indexes instead

Maxxen committed:
fix tidy warnings

Maxxen committed:
add include

Maxxen committed:
Merge branch 'main' into custom-index-fork

Maxxen committed:
more initialization, another include

Maxxen committed:
more index initialization, serialization

Maxxen committed:
move scan initialization into index

Maxxen committed:
Merge branch 'main' into custom-index-fork

Maxxen commented:
@Mytherin The regression failure seems unrelated, could you rerun?

taniabogatsch commented on src/execution/index/art/art.cpp:
```diff
@@ -804,7 +894,7 @@ bool ART::SearchCloseRange(ARTIndexScanState &state, ARTKey &lower_bound, ARTKey
 }
 
-bool ART::Scan(const Transaction &transaction, const DataTable &table, IndexScanState &state, const idx_t max_count,
+bool ART::Scan(const Transaction &transaction, const DataTable &table, IndexScanState &state, idx_t max_count,
```
Why is this no longer `const`?

Maxxen replied:
It barely has any effect being const and I get CLion warnings by having it be const in the header as it has no effect there - but I can change it back.

taniabogatsch commented on src/storage/table_index_list.cpp:
```diff
@@ -50,6 +53,34 @@ bool TableIndexList::NameIsUnique(const string &name) {
+			auto &create_info = unknown_index.GetCreateInfo();
+			auto &storage_info = unknown_index.GetStorageInfo();
+
+			auto index_instance = index_type_entry->create_instance(
+			    create_info.index_name, create_info.constraint_type, create_info.column_ids,
+			    unknown_index.unbound_expressions, *table_info.table_io_manager, table_info.db, storage_info);
```
Can we run this callback function on the index instead of using the catalog entry?

Maxxen replied:
You mean pass it the unknown index? We don't always have one (e.g when creating a new index). I've been thinking of wrapping these args together into a struct thought to make it less cruddy.

My hypothesis is that we're gonna end up with more function pointers here too.

taniabogatsch commented on src/storage/checkpoint_manager.cpp:
```diff
@@ -451,9 +465,18 @@ void CheckpointReader::ReadIndex(ClientContext &context, Deserializer &deseriali
-	                          std::move(unbound_expressions), data_table.db, nullptr, index_storage_info);
-	data_table.info->indexes.AddIndex(std::move(art));
+
+	if (info.index_type == "ART") {
+		data_table.info->indexes.AddIndex(make_uniq<ART>(info.index_name, info.constraint_type, info.column_ids,
+		                                                 TableIOManager::Get(data_table), unbound_expressions,
+		                                                 data_table.db, nullptr, index_storage_info));
+	} else {
+		auto unknown_index = make_uniq<UnknownIndex>(info.index_name, info.index_type, info.constraint_type,
+		                                             info.column_ids, TableIOManager::Get(data_table),
+		                                             unbound_expressions, data_table.db, info, index_storage_info);
+
+		data_table.info->indexes.AddIndex(std::move(unknown_index));
+	}
```
We can unify this. The ART can still be an unknown index at this point, and we can initialize it the same way we're initializing other indexes later.

Maxxen replied:
The ART index type is always setup by the default catalog generator, so AFAIK it can't be uninitialized here.

taniabogatsch commented on src/storage/local_storage.cpp:
```diff
@@ -23,7 +23,12 @@ LocalTableStorage::LocalTableStorage(DataTable &table)
 
 	table.info->indexes.Scan([&](Index &index) {
+		if (index.index_type != "ART") {
+			// TODO: support other index types
```
Nit.

taniabogatsch reviewed (changes requested):
Hey @Maxxen, thanks for the PR! Below are my comments.

- we're adding a lot of new code here; can we run the code coverage on it?

We have to reconsider how we link catalog index types to the storage... Can we move the callback functions into a `DuckIndexType`? Then, we can treat the ART the same as other unknown indexes and only initialize it, if necessary, as part of duck table operations.

Maxxen commented on src/catalog/catalog_entry/schema_catalog_entry.cpp:
```diff
@@ -43,4 +43,9 @@ string SchemaCatalogEntry::ToSQL() const {
 	return create_schema_info->ToString();
 }
 
+optional_ptr<CatalogEntry> SchemaCatalogEntry::CreateIndexType(CatalogTransaction transaction,
+                                                               CreateIndexTypeInfo &info) {
+	throw NotImplementedException("Schema does not support creating index types!");
+}
```
This is the base schema class, that potentially different catalog types could subclass, e.g the postgres extensions provide `PostgresSchemaCatalogEntry`. But It doesn't make sense to add a custom extension index type to a non-duckdb catalog, therefore we just throw here by default. The DuckSchemaEntry overrides this.

Maxxen commented on src/include/duckdb/catalog/catalog_entry/index_type_catalog_entry.hpp:
```diff
@@ -0,0 +1,39 @@
+
+#include "duckdb/catalog/standard_entry.hpp"
+#include "duckdb/parser/parsed_data/create_index_type_info.hpp"
+#include "duckdb/storage/index.hpp"
```
We only support custom index types for DuckCatalog anyway, so there is no point having a separate `DuckIndexTypeEntry` compared to `IndexTypeEntry`. 

It doesn't really make sense to add an index type from an extension to a non-duckdb storage. (we don't have or want to have cross-extension dependencies).

Maxxen commented on src/include/duckdb/parser/parsed_data/create_index_type_info.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace duckdb {
+
+struct CreateIndexTypeInfo : public CreateInfo {
+	explicit CreateIndexTypeInfo(string name_p, string schema = DEFAULT_SCHEMA);
```
We just default to the `DEFAULT_SCHEMA`. Extensions can still add internal entries. We need it to be internal so that we don't write it to the WAL.

Maxxen commented on src/catalog/default/default_index_types.cpp:
```diff
@@ -0,0 +1,55 @@
+	auto result = make_uniq_base<CatalogEntry, IndexTypeCatalogEntry>(catalog, schema, info);
+
+	auto &index_entry = result->Cast<IndexTypeCatalogEntry>();
+	index_entry.create_instance =
+	    [](const string &name, const IndexConstraintType constraint_type, const vector<column_t> &column_ids,
+	       const vector<unique_ptr<Expression>> &unbound_expressions, TableIOManager &table_io_manager,
+	       AttachedDatabase &db, const IndexStorageInfo &storage_info) -> unique_ptr<Index> {
+		auto art = make_uniq<ART>(name, constraint_type, column_ids, table_io_manager, unbound_expressions, db, nullptr,
+		                          storage_info);
+		return std::move(art);
+	};
```
I elaborated on this above, but I don't think it makes sense to have a separate DuckIndexType? I could create one anyway to be consistent/future proof though.

Maxxen replied:
Ahh maybe it make sense for MD.

Maxxen commented:
> * we're adding a lot of new code here; can we run the code coverage on it?

As soon as I can get an example running the whole lifecycle of a custom index I can add it as a cpp test maybe.

> We have to reconsider how we link catalog index types to the storage... Can we move the callback functions into a `DuckIndexType`? Then, we can treat the ART the same as other unknown indexes and only initialize it, if necessary, as part of duck table operations.

Sure, I can add a DuckIndexType. But we I don't think we need to worry about the ART being uninitialized as I added it to the default catalog generator. e.g. its catalog entry is always lazily added when referenced.

Maxxen committed:
small refactoring

Maxxen committed:
merge

Mytherin commented on src/include/duckdb/catalog/catalog_entry/index_type_catalog_entry.hpp:
```diff
@@ -0,0 +1,39 @@
+                                                     const IndexStorageInfo &storage_info);
+
+//! An catalog entry for an index "type"
+class IndexTypeCatalogEntry : public StandardEntry {
```
Perhaps we can move this to the `DBConfig`?

Mytherin commented on src/storage/table_index_list.cpp:
```diff
@@ -50,6 +53,34 @@ bool TableIndexList::NameIsUnique(const string &name) {
+void TableIndexList::InitializeIndexes(ClientContext &context, DataTableInfo &table_info) {
+	lock_guard<mutex> lock(indexes_lock);
+	for (auto &index : indexes) {
+		if (index->IsUnknown()) {
```
nit: early-out

Maxxen committed:
switch from catalog entry to dbconfig

Maxxen committed:
add header

Maxxen committed:
format and delete unused file

Maxxen commented:
@taniabogatsch After considering your feedback and also talking to @Mytherin I've changed the index "type" to instead live in the DbConfig instead of the catalog. IndexType's don't really have anything to store anyway and since they are always pushed by extensions (or built in) don't really need transactional support either.

This PR also moved some of the index scan decision code into the ART class itself instead of being part of the table scan. In my custom index extension I've realized that its probably easier to just swap out the table scan function for a separate custom index scan with an optimizer rule so this change is not strictly necessary to achieve the goal I'm working towards right now but I think keeping it in is a good idea and a step in the right direction if we want to add a proper abstraction for index access methods in the future.

I've also moved the `"ART"` magic string to instead be a static constexpr member of the ART class and moved the "create_instance" function for the art type to point to to a static member method in the ART class instead of a lambda. Also some other helper classes like `IndexTypeSet`, but overall the footprint of this PR is now quite a bit smaller.

Maxxen committed:
tidy

Mytherin commented:
Thanks!
